HDDS-5534. Verify config is updated on all OMs before proceeding with Bootstrap

hanishakoneru opened the pull request:
## What changes were proposed in this pull request?

When an OM is bootstrapped, it sends a SetConfiguration ratis request to leader OM and this request is propagated to all existing OMs in the ring. Existing OMs add the new OM to their peer list by reloading the config files and getting the new node information from the reloaded config files. But if the config file (ozone-site.xml) has not been updated on any OM, then that OM would crash when processing the SetConfiguration request. 

To avoid this scenario, before a bootstrap request is sent to the leader OM, the bootstrapping OM must verify that all the existing OMs have its information in their ozone-site.xml. To achieve this, we have added a new OMMetadata protocol. This protocol would be called during OM decommissioning also to verify that ozone-site.xmls are updated.

It could happen that an OM is down or not responding to a bootstrapping OMs metadata information call. In this scenario, to proceed with bootstrap, a "force" option is provided. During force bootstrap, the check for config update on existing OMs will be skipped.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5534

## How was this patch tested?

Will add unit tests

hanishakoneru commented:
@bharatviswa504 can you please take a look when you can. Thanks.

bharatviswa504 reviewed (commented):
Started review, due to other tasks not able to complete review.

One question what advantage of this validation compared to failing with exception, will we be in some inconsistent state? I just want to understand reason behind going this route. Because if any of the node is down, we cannot make use of this check i believe. I am still going through patch BTW.

hanishakoneru commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OMNodeDetails.java:
```diff
@@ -164,39 +166,80 @@ public String getOMDBCheckpointEnpointUrl(boolean isHttpPolicy) {
+    String hostAddr;
+    InetSocketAddress omRpcAddress = null;
+    int rpcPort = 0;
+    if (shouldResolveAddr) {
```
I was trying to avoid creating a socket address when we only need to host and port string. But yes, we can throw an exception too. Though it would be run on all the OMs, the overhead will be minimal.

hanishakoneru commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocol/OMMetadata.java:
```diff
@@ -0,0 +1,77 @@
+ * information through the {@link OMMetadataProtocol} and not for storing the
+ * metadata information in OzoneManager itself.
+ */
+public final class OMMetadata {
```
We have already defined OMNodeDetails and OMNodeInfo. What do you think of OMConfiguration (similar to RaftConfiguration)?

hanishakoneru commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocol/OMMetadataProtocol.java:
```diff
@@ -0,0 +1,36 @@
+ */
+@KerberosInfo(
+    serverPrincipal = OMConfigKeys.OZONE_OM_KERBEROS_PRINCIPAL_KEY)
+public interface OMMetadataProtocol extends Closeable {
```
In future, if there is a need, this protocol can be reused for other meta information about the OM.

How about renaming the RPC request instead from `getOMMetadataInfo` to `getOMConfiguration`?

bharatviswa504 replied:
Just a question, why cant we add this to OMInterServiceProtocol?
Any reason to add it as a fresh OMMetadataProtocol?

hanishakoneru commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OMMetadataProtocolClientSideImpl.java:
```diff
@@ -0,0 +1,128 @@
+    // OM metadata is requested from a specific OM and hence there is no need
+    // of any failover provider.
+    RetryPolicy connectionRetryPolicy = RetryPolicies
+        .failoverOnNetworkException(0);
```
Done.

hanishakoneru commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OMMetadataProtocolClientSideImpl.java:
```diff
@@ -0,0 +1,128 @@
+      throws IOException {
+
+    RPC.setProtocolEngine(OzoneConfiguration.of(conf),
+        OMInterServiceProtocolPB.class, ProtobufRpcEngine.class);
```
Done.

hanishakoneru commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocol/OMMetadata.java:
```diff
@@ -0,0 +1,77 @@
+
+  private OMMetadata(List<OMNodeDetails> inMemoryNodeList,
+      List<OMNodeDetails> onDiskNodeList) {
+    if (inMemoryNodeList != null) {
```
Done.

hanishakoneru commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java:
```diff
@@ -697,4 +701,51 @@ public static String normalizeKey(String keyName,
+    int rpcPort, ratisPort;
+    if (omNodeIds.size() == 0) {
+      //Check if it is non-HA cluster
+      rpcAddrKey = ConfUtils.addKeySuffixes(OZONE_OM_ADDRESS_KEY);
```
Done.

hanishakoneru commented:
Thank you @bharatviswa504 for reviewing the PR. I have addressed some of your comments and replied to the rest.

> One question what advantage of this validation compared to failing with exception, will we be in some inconsistent state? I just want to understand reason behind going this route. Because if any of the node is down, we cannot make use of this check i believe.

If the config has not been updated on an OM and the SetConfiguration request is executed, then it could lead to that OM crashing out. If we perform this check before, we can avoid a healthy OM going down because of not updating its config with the new OM info. A healthy running OM should not crash when trying to add another OM.

If a node is down, this check would at the very least warn the user to double check the configs before proceeding with force bootstapping.

hanishakoneru commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -1373,6 +1411,64 @@ public void restart() throws IOException {
+  private boolean checkOMexistsInRemoteOMConfig(OMMetadata remoteOMMetadata) {
+    List<OMNodeDetails> omNodesInNewConf = remoteOMMetadata
+        .getOmNodesInNewConf();
+    for (OMNodeDetails omNodeInRemoteOM : omNodesInNewConf) {
```
There are 2 extra checks which I wanted to add but forgot.
1. New OM is not already present as part of the Ratis ring
2. A nodeId is not being reused.
Also, we can use this method when removing an OM to verify that the OMs are in sync.
But I see your point. Let me test what the implications would be for the 2 scenarios mentioned above. If its harmless and results in the bootstrapping node to shutdown, we should be fine. Otherwise, I will add those checks.

hanishakoneru replied:
I tested on a cluster and verified that bootstrapping a node again would lead it to crash. And the exception comes from Ratis. Instead, we should keep this check here to avoid that.

hanishakoneru commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OMMetadataProtocolServerSideImpl.java:
```diff
@@ -0,0 +1,66 @@
+    List<OMNodeInfo> omNodesInNewConf =
+        new ArrayList<>(newOMNodesList.size());
+    for (OMNodeDetails omNodeDetails : newOMNodesList) {
+      omNodesInNewConf.add(omNodeDetails.getProtobuf());
```
Neither getAllOMNodesInMemory() nor getAllOMNodesInNewConf() return null. They return an empty arraylist when there are no elements. So it should be ok as we will not be adding a null value to proto.

hanishakoneru commented:
Thank you @bharatviswa504 for reviewing this and for the comments. I have addressed them and made some improvements as discussed above in the latest patch. I am working on unit tests and will post an update soon.

hanishakoneru commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -1373,6 +1411,64 @@ public void restart() throws IOException {
+   * configs.
+   */
+  private boolean checkOMexistsInRemoteOMConfig(OMMetadata remoteOMMetadata) {
+    List<OMNodeDetails> omNodesInNewConf = remoteOMMetadata
```
Done.

hanishakoneru commented:
@bharatviswa504 I have added some unit tests and made improvements in the existing bootstrap tests as well. Please take a look when you get a chance. Thanks.

bharatviswa504 commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocol/OMConfiguration.java:
```diff
@@ -0,0 +1,84 @@
+    }
+  }
+
+  public List<String> getCurrentPeerList() {
```
Missing JavaDocs

hanishakoneru replied:
Done.

bharatviswa504 commented on hadoop-ozone/interface-client/src/main/proto/OMMetadataProtocol.proto:
```diff
@@ -0,0 +1,65 @@
+    required string nodeID = 1;
+    required string hostAddress = 2;
+    required uint32 rpcPort = 3;
+    required uint32 ratisPort = 4;
```
Any reason for adding ratisport, I see it is not checked?
And also we donot check about httpaddress? 

Is the intention here is to check if already started OM has this new OM address, and OM Ratis port only. Because if http address is not there during bootstrap we will fail correct?

hanishakoneru replied:
We do not check for either ratis port match or http address match. I can add checks for these two or remove ratisPort also from here. Let me know what you think.

hanishakoneru replied:
We currently do not check for ratisPort and httpAddr match while doing bootstrap check. I can add them if you think it is necessary.

bharatviswa504 replied:
@hanishakoneru As discussed offline, we can do this a new Jira and also comparing remote OM config with local config to check if the new OM has all the older OM config. Pls open a new Jira for this.

bharatviswa504 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -1453,7 +1559,13 @@ public void updatePeerList(List<String> omNodeIds) {
+        try {
+          addOMNodeToPeers(omNodeId);
+        } catch (IOException e) {
+          LOG.error("Fatal Error: Shutting down the system as otherwise it " +
```
Question to understand: Why OM's will diverge? And also why are we shutting down OM here?

From PR description, i thought if config is not updated on existing OM's we are shutting down, but shutdown if new config does not have that node id is being done as part of this Jira.

hanishakoneru replied:
This function is called from the StateMachine. Any error in SM could be fatal as it diverges the OM states. Let's say  om1 and om2 update their configs about new nodes om4 and om5, but om3 does not as addOMNodeToPeers fails. This will lead to om3 being in a different config without having knowledge about om4 and om5. Now let's say om1 and om2 go down and om4 is the new leader. om3 cannot contact om4 to download a checkpoint if required as it does not know the addresses of om4.

hanishakoneru replied:
This function is called from OM SM. Any failure here could be fatal as OM states would diverge.
For example, let's say OM1 and OM2 successfully update their configs with newly bootstrapped nodes OM4 and OM5. OM3 fails to do so. Let's say later OM1 and OM2 shutdown with OM4 as the new leader. OM3 would not be able to download checkpoint from OM4 as it is not aware of OM4 addresses.

bharatviswa504 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOzoneManagerBootstrap.java:
```diff
@@ -205,12 +191,146 @@ public void testLeaderChangeToNewOM() throws Exception {
+    }
+
+    /***************************************************************************
+     * 2. Force bootstrap without upating config on any OM -> fail
```
Minor NIT: upating -> updating

hanishakoneru replied:
Done.

bharatviswa504 commented:
Thank You @hanishakoneru for the work.
Overall approach LGTM, few comments/questions

hanishakoneru commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java:
```diff
@@ -697,4 +701,78 @@ public static String normalizeKey(String keyName,
+    return omNodesList;
+  }
+
+  public static String getOMAddressListPrintString(List<OMNodeDetails> omList) {
```
Done.

hanishakoneru replied:
Done.

hanishakoneru commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OMMetadataProtocolClientSideImpl.java:
```diff
@@ -0,0 +1,136 @@
+      }
+      return omMedatataBuilder.build();
+    } catch (ServiceException e) {
+      LOG.error("Failed to retrieve configuration of OM {}", omNodeID, e);
```
Done.

hanishakoneru commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OMMetadataProtocolServerSideImpl.java:
```diff
@@ -0,0 +1,66 @@
+
+/**
+ * This class is the server-side translator that forwards requests received on
+ * {@link OMMetadataProtocolPB} to the OMConfiguration server implementation.
```
Done.

hanishakoneru commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -1648,7 +1779,7 @@ public long getRatisSnapshotIndex() throws IOException {
    */
   public void stop() {
-    LOG.info("Stopping Ozone Manager");
+    LOG.info("{}: Stopping Ozone Manager", getOMNodeId());
```
Done.

hanishakoneru commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -1531,6 +1650,18 @@ public boolean doesPeerExist(String omNodeId) {
     return false;
   }
 
+  public List<OMNodeDetails> getAllOMNodesInMemory() {
```
Done.

hanishakoneru commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -1531,6 +1650,18 @@ public boolean doesPeerExist(String omNodeId) {
+    return peerNodes;
+  }
+
+  public List<OMNodeDetails> getAllOMNodesInNewConf() {
```
Done.

hanishakoneru commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -1423,6 +1466,66 @@ public void restart() throws IOException {
+
+    if (remoteOMConfig.getCurrentPeerList().contains(this.getOMNodeId())) {
+      throw new IOException("Remote OM " + remoteNodeId + " already contains " +
+          "bootstrapping OM(" + getOMNodeId() + ") in it's in memory peer " +
```
Done.

hanishakoneru commented:
Thanks @bharatviswa504 for the review. I have addressed your comments in the new patch or answered inline. 

> Just a question, why cant we add this to OMInterServiceProtocol?
Any reason to add it as a fresh OMMetadataProtocol?
Added a new protocol so that it can be used for removeOM admin command also. RemoveOM would have to be an admin command and not from other OMs. So we need a new protocol for admins to perform operations on the OM.

bharatviswa504 commented:
> Thanks @bharatviswa504 for the review. I have addressed your comments in the new patch or answered inline.
> 
> > Just a question, why cant we add this to OMInterServiceProtocol?
> > Any reason to add it as a fresh OMMetadataProtocol?
> 
> Added a new protocol so that it can be used for removeOM admin command also. RemoveOM would have to be an admin command and not from other OMs. So we need a new protocol for admins to perform operations on the OM.

As this verify configuration is InterOM only right, not external client interaction. And we can add to OmClientProtocol that API right, do we need a new protocol for admin operations. And in OM server we can check whether admin has issued the command or not correct, Let me know if i am missing something here.

bharatviswa504 commented:
Thank You @hanishakoneru for the update.
Overall LGTM, one question on the new protocol thing, commented inline.

hanishakoneru commented:
> As this verify configuration is InterOM only right, not external client interaction. And we can add to OmClientProtocol that API right, do we need a new protocol for admin operations. And in OM server we can check whether admin has issued the command or not correct, Let me know if i am missing something here.

It would be better to keep the admin protocols separate from client protocols. RemoveOM operation is a critical operation and I don't think it should be exposed to clients at all.

bharatviswa504 commented:
> > As this verify configuration is InterOM only right, not external client interaction. And we can add to OmClientProtocol that API right, do we need a new protocol for admin operations. And in OM server we can check whether admin has issued the command or not correct, Let me know if i am missing something here.
> 
> It would be better to keep the admin protocols separate from client protocols. RemoveOM operation is a critical operation and I don't think it should be exposed to clients at all.

But the OmMetadatProtocol, is used between inter OM's if i understand correctly, the admins/clients will not interact with this.
What is the advantage of the seperating admin protocol, as when it comes to implementation from client this will be exposed from admin command like ozone admin, and we check at server if user is admin or not. 
Do we see any benifits of having a seperate protocol for admin operations? And also as said this protocol which is added is again used between OM's, so I donot understand of linking between admin here.

bharatviswa504 commented:
From security perspective one I can think of is, we can add service level ACL's so that it can be rejected by Rpc Layer it self.
But not sure this is that critical, as if we have checks for admin methods in server, the same can be achieved.

bharatviswa504 commented:
>Thanks @bharatviswa504 for the review.
>I have opened HDDS-5978 to add more checks to the bootstrap process.

I have moved this Jira under HDDS-4330.

hanishakoneru commented:
@bharatviswa504 thanks for the offline discussion. I have updated to Metadata protocol name to Admin Protocol to better represent the purpose.

bharatviswa504 commented on hadoop-ozone/interface-client/src/main/proto/OMAdminProtocol.proto:
```diff
@@ -57,7 +57,7 @@ message OMNodeInfo {
 /**
  The service for getting OM metadata.
```
Minor: Can we change doc to have something like service for OM admin operations??

bharatviswa504 reviewed (commented):
LGTM. Few minor comments.

hanishakoneru commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestOzoneConfigurationFields.java:
```diff
@@ -100,7 +100,9 @@ private void addPropertiesNotInXml() {
         OMConfigKeys.OZONE_OM_RATIS_SNAPSHOT_AUTO_TRIGGER_THRESHOLD_KEY,
-        OMConfigKeys.OZONE_OM_HA_PREFIX
+        OMConfigKeys.OZONE_OM_HA_PREFIX,
+        OMConfigKeys.OZONE_OM_ADMIN_PROTOCOL_MAX_RETRIES_KEY,
```
As these are just used for admin operations, I don't think tuning these would make much of a difference. It would still be configurable but just not exposed via ozone-default.xml.

bharatviswa504 replied:
I think it would be good to add in ozone-default.xml.
So that users knows about these configs, otherwise now they need to know only by reading code.

bharatviswa504 replied:
Because atleast internally I have seen a bug for admin commands, where using defaults it is taking longer, the users needs some settings to control retry behavior. (This is for SCM, and in SCM admin commands are part of normal client/block protocol, as in OM we have separated it out that is good.)

hanishakoneru commented:
Thanks @bharatviswa504. I have addressed the comments.

bharatviswa504 reviewed (approved):
+1 LGTM.
Thank You @hanishakoneru for diligently working on this PR.

bharatviswa504 commented:
Test failures looks unrelated, triggered another CI run.

hanishakoneru commented:
@bharatviswa504, thanks for reviewing multiple times and approving it. I will merge this shortly.
